pkg/apiserver/internal/v1/offering.go

```diff
@@ -0,0 +1,146 @@
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	catalogv1alpha1 "github.com/kubermatic/kubecarrier/pkg/apis/catalog/v1alpha1"
+	v1 "github.com/kubermatic/kubecarrier/pkg/apiserver/api/v1"
+)
+
+type offeringServer struct {
+	client client.Client
+}
+
+var _ v1.OfferingServiceServer = (*offeringServer)(nil)
+
+// +kubebuilder:rbac:groups=catalog.kubecarrier.io,resources=offerings,verbs=get;list
+
+func NewOfferingServiceServer(c client.Client) v1.OfferingServiceServer {
+	return &offeringServer{
+		client: c,
+	}
+}
+
+func (o offeringServer) List(ctx context.Context, req *v1.OfferingListRequest) (res *v1.OfferingList, err error) {
+	var listOptions []client.ListOption
+	listOptions, err = o.validateListRequest(req)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, err.Error())
+	}
+	offeringList := &catalogv1alpha1.OfferingList{}
+	if err := o.client.List(ctx, offeringList, listOptions...); err != nil {
+		return nil, status.Errorf(codes.Internal, fmt.Sprintf("listing offerings: %s", err.Error()))
+	}
+
+	res = &v1.OfferingList{}
+	for _, catalogOffering := range offeringList.Items {
+		res.Items = append(res.Items, o.convertOffering(&catalogOffering))
+	}
+	res.Continue = offeringList.Continue
+	return
+}
+
+func (o offeringServer) Get(ctx context.Context, req *v1.OfferingRequest) (res *v1.Offering, err error) {
+	if err = o.validateGetRequest(req); err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, err.Error())
+	}
+	offering := &catalogv1alpha1.Offering{}
+	if err = o.client.Get(ctx, types.NamespacedName{
+		Name:      req.Name,
+		Namespace: req.Tenant,
+	}, offering); err != nil {
+		return nil, status.Errorf(codes.Internal, fmt.Sprintf("getting offering: %s", err.Error()))
+	}
+	return o.convertOffering(offering), nil
+
+}
+
+func (o offeringServer) validateListRequest(req *v1.OfferingListRequest) ([]client.ListOption, error) {
+	var listOptions []client.ListOption
+	if req.Tenant == "" {
+		return listOptions, fmt.Errorf("missing namespace")
+	}
+	listOptions = append(listOptions, client.InNamespace(req.Tenant))
+	if req.Limit < 0 {
+		return listOptions, fmt.Errorf("invalid limit: should not be negative number")
+	}
+	listOptions = append(listOptions, client.Limit(req.Limit))
+	if req.LabelSelector != "" {
+		selector, err := labels.Parse(req.LabelSelector)
+		if err != nil {
+			return listOptions, fmt.Errorf("invalid LabelSelector: %w", err)
+		}
+		listOptions = append(listOptions, client.MatchingLabelsSelector{
+			Selector: selector,
+		})
+	}
+	if req.Continue != "" {
+		listOptions = append(listOptions, client.Continue(req.Continue))
+	}
+	return listOptions, nil
+}
+
+func (o offeringServer) validateGetRequest(req *v1.OfferingRequest) error {
```

Comment:
because we are putting all service in the same `v1` package it's better to name this method something like `validateOfferingRequest`, in that case we can use it in other services that need to use `offering`

Reply:
1. the receiver of this function is `offeringServer` so naming it to `validateOfferingRequest` is a bit redundant.
> in that case we can use it in other services that need to use offering
2. I don't really understand this, why other services will use `OfferingRequest`? Can you please provide an example here?

Reply:
> 1. the receiver of this function is `offeringServer` so naming it to `validateOfferingRequest` is a bit redundant.

yeah, I was thinking about decoupling those methods from `offeringServer` because they don't really need it at all 

but, meh, I think you are right, let's make it simple